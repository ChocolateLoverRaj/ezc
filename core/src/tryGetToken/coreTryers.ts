import CoreTokenWithData from '../CoreTokenWithData'
import parseIdentifier from './parseIdentifier'
import parseIntegerType from './parseIntegerType'
import coreInput from './parseKeyword/coreInput'
import parseKeyword from './parseKeyword/parseKeyword'
import parseNumberLiteral from './parseNumberLiteral'
import parseStringLiteral from './parseStringLiteral'
import TryGetToken from './TryGetToken'

const coreTryers: ReadonlyArray<TryGetToken<CoreTokenWithData>> = [
  parseIntegerType,
  parseKeyword(coreInput),
  parseStringLiteral,
  parseIdentifier,
  parseNumberLiteral
] as const

export default coreTryers
