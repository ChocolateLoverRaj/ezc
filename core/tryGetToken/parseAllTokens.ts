import skip from '../../util/splitAsyncIterator/skip'
import splitAsyncIterator from '../../util/splitAsyncIterator/splitAsyncIterator'
import EnumItemWithData from '../EnumItemWithData'
import isStreamWhitespace from '../../util/isStreamWhitespace/isStreamWhitespace'
import parseToken from './parseToken'
import TryGetToken from './TryGetToken'

const parseAllTokens = (tryers: ReadonlyArray<TryGetToken<EnumItemWithData>>) => async function * (
  asyncIterable: AsyncIterable<string>
): AsyncGenerator<EnumItemWithData | undefined> {
  const splittedIterator = splitAsyncIterator(asyncIterable[Symbol.asyncIterator]())
  let chunkBefore = ''
  while (true) {
    const asyncIterable: AsyncIterable<string> = {
      async * [Symbol.asyncIterator] () {
        yield chunkBefore
        for await (const chunk of splittedIterator.asyncIterable) {
          yield chunk
        }
      }
    }
    const token = await parseToken(tryers)(asyncIterable)
    if (token === undefined) {
      // If the string wasn't empty don't yield undefined because it means something couldn't be
      // parsed
      if (!await isStreamWhitespace(asyncIterable)) yield undefined
      return
    }
    yield token.token
    let chunksProcessed = 0
    let charactersProcessed = 0
    if (token.length < chunkBefore.length) {
      chunkBefore = chunkBefore.slice(token.length - chunkBefore.length)
    } else {
      for await (const chunk of splittedIterator.asyncIterable) {
        chunksProcessed++
        charactersProcessed += chunk.length
        if (charactersProcessed >= token.length) {
          chunkBefore = charactersProcessed === token.length
            ? ''
            : chunk.slice(token.length - charactersProcessed)
          break
        }
      }
      if (chunksProcessed === 0) return
      skip(splittedIterator, chunksProcessed)
    }
  }
}

export default parseAllTokens
