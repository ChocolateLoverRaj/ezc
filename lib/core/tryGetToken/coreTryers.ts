import CoreTokenWithData from '../CoreTokenWithData'
import coreParseKeywordOptions from './coreParseKeywordOptions'
import parseIdentifier from './parseIdentifier'
import parseIntegerType from './parseIntegerType'
import parseKeyword from './parseKeyword'
import parseNumberLiteral from './parseNumberLiteral'
import parseOpenClose from './parseOpenClose'
import parseStringLiteral from './parseStringLiteral'
import TryGetToken from './TryGetToken'

const coreTryers: ReadonlyArray<TryGetToken<CoreTokenWithData>> = [
  parseIdentifier,
  parseIntegerType,
  parseKeyword(coreParseKeywordOptions),
  parseNumberLiteral,
  parseOpenClose,
  parseStringLiteral
] as const

export default coreTryers
