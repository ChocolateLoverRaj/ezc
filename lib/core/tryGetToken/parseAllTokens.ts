import EnumItemWithData from '../EnumItemWithData'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import parseToken from './parseToken'
import TryGetToken from './TryGetToken'

const parseAllTokens = (tryers: ReadonlyArray<TryGetToken<EnumItemWithData>>) => async function * (
  asyncIterable: AsyncIterable<string>
): AsyncGenerator<EnumItemWithData | undefined> {
  const splittedIterator = splitAsyncIterator(asyncIterable[Symbol.asyncIterator]())
  let chunkBefore = ''
  while (true) {
    const token = await parseToken(tryers)({
      async * [Symbol.asyncIterator] () {
        yield chunkBefore
        for await (const chunk of splittedIterator.asyncIterable) {
          yield chunk
        }
      }
    })
    if (token === undefined) {
      yield undefined
      return
    }
    yield token.token
    let chunksProcessed = 0
    let charactersProcessed = 0
    if (token.length < chunkBefore.length) {
      chunkBefore = chunkBefore.slice(token.length - chunkBefore.length)
    } else {
      for await (const chunk of splittedIterator.asyncIterable) {
        chunksProcessed++
        charactersProcessed += chunk.length
        if (charactersProcessed >= token.length) {
          chunkBefore = charactersProcessed === token.length
            ? ''
            : chunk.slice(token.length - charactersProcessed)
          break
        }
      }
      if (chunksProcessed === 0) return
      skip(splittedIterator, chunksProcessed)
    }
  }
}

export default parseAllTokens
