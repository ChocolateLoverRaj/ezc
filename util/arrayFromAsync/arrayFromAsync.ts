const arrayFromAsync = async <T>(asyncIterator: AsyncIterator<T>): Promise<T[]> => {
  const arr: T[] = []
  while (true) {
    const { value, done } = await asyncIterator.next()
    if (done === true) return arr
    arr.push(value)
  }
}

export default arrayFromAsync
