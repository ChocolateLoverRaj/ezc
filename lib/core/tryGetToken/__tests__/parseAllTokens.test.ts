import arrayFromAsync from '../../arrayFromAsync'
import coreTryers from '../coreTryers'
import parseAllTokens from '../parseAllTokens'

test('declare i32 @puts(ptr)', async () => {
  await expect(arrayFromAsync(parseAllTokens(coreTryers)({
    async * [Symbol.asyncIterator] () {
      yield 'declare i32 @puts(ptr)'
    }
  }))).resolves.toMatchSnapshot()
})
