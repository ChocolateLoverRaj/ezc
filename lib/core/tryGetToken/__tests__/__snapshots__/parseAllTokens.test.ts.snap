// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`( 1`] = `
Array [
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 19,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
]
`;

exports[`(, 1`] = `
Array [
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 19,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 15,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
]
`;

exports[`declare i32 @puts(ptr) 1`] = `
Array [
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 4,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
  Object {
    "data": 32,
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 2,
    },
  },
  Object {
    "data": Object {
      "name": "puts",
      "type": 0,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 0,
    },
  },
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 19,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 6,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
  Object {
    "data": Object {
      "enum": Object {
        "0": "PRIVATE",
        "1": "UNNAMED_ADDR",
        "10": "RETURN",
        "11": "EQUALS",
        "12": "GLOBAL",
        "13": "NO_CAPTURE",
        "14": "NO_ALIAS",
        "15": "COMMA",
        "16": "VOID",
        "17": "NO_FREE",
        "18": "CALL",
        "19": "OPEN_PARENTHESIS",
        "2": "CONSTANT",
        "20": "CLOSE_PARENTHESIS",
        "21": "OPEN_BRACKET",
        "22": "CLOSE_BRACKET",
        "23": "OPEN_CURLY_BRACKET",
        "24": "CLOSE_CURLY_BRACKET",
        "3": "ALIGN",
        "4": "DECLARE",
        "5": "DEFINE",
        "6": "PTR",
        "7": "GET_ELEMENT_PTR",
        "8": "INBOUNDS",
        "9": "X",
        "ALIGN": 3,
        "CALL": 18,
        "CLOSE_BRACKET": 22,
        "CLOSE_CURLY_BRACKET": 24,
        "CLOSE_PARENTHESIS": 20,
        "COMMA": 15,
        "CONSTANT": 2,
        "DECLARE": 4,
        "DEFINE": 5,
        "EQUALS": 11,
        "GET_ELEMENT_PTR": 7,
        "GLOBAL": 12,
        "INBOUNDS": 8,
        "NO_ALIAS": 14,
        "NO_CAPTURE": 13,
        "NO_FREE": 17,
        "OPEN_BRACKET": 21,
        "OPEN_CURLY_BRACKET": 23,
        "OPEN_PARENTHESIS": 19,
        "PRIVATE": 0,
        "PTR": 6,
        "RETURN": 10,
        "UNNAMED_ADDR": 1,
        "VOID": 16,
        "X": 9,
      },
      "id": 20,
    },
    "type": Object {
      "enum": Object {
        "0": "IDENTIFIER",
        "1": "NUMBER_LITERAL",
        "2": "INTEGER_TYPE",
        "3": "STRING_LITERAL",
        "4": "ENTRY_BLOCK",
        "5": "KEY_WORD",
        "ENTRY_BLOCK": 4,
        "IDENTIFIER": 0,
        "INTEGER_TYPE": 2,
        "KEY_WORD": 5,
        "NUMBER_LITERAL": 1,
        "STRING_LITERAL": 3,
      },
      "id": 5,
    },
  },
]
`;
